File: Heap/RemoveDuplicateLetter/RemoveDuplicateLetters.java
import java.util.Stack;
import java.util.Arrays;
import java.lang.StringBuilder;

public class RemoveDuplicateLetters{
	public static String removeDuplicateLetters(String s){
		if(s.length()==0 || s==null) return "";
		int[] occurrence=new int[26];	//count # of occurrences of each letter
		boolean[] visited=new boolean[26];	//mark if visited of each letter
		char[] ch=s.toCharArray();
		for(char c:ch)
			occurrence[c-'a']++;
		//System.out.println(Arrays.toString(occurrence));
		int idx=0;
		Stack<Character> stack=new Stack<Character>();

		for(char c:ch){
			idx=c-'a';
			occurrence[idx]--;
			if(!visited[idx]){	//if current letter hasn't appeared yet
				while(!stack.isEmpty() && c<stack.peek() && occurrence[stack.peek()-'a']!=0){
					char oldTop=stack.pop();
					visited[oldTop-'a']=false;
				}
				stack.push(c);
				visited[idx]=true;
			}
		}

		StringBuilder res=new StringBuilder();
		while(!stack.isEmpty())
			res.append(stack.pop());
		return res.reverse().toString();
	}

	public static void main(String[] args) {
		String s="cbacdcbc";
		System.out.println("'"+s+"' -> '"+removeDuplicateLetters(s)+"'.");
	}
}

File: Python/RemoveDuplicateLetters.py
class RemoveDuplicateLetters(object):
	def removeDuplicateLetters(self,s):
		string=""
		stack=[]
		visited=[0]*26
		appeared=[0]*26
		for c in s:
			appeared[ord(c)-ord('a')]+=1

		for c in s:
			i=ord(c)-ord('a')
			appeared[i]-=1
			if visited[i]==0:
				while len(stack)>0 and appeared[ord(stack[-1])-ord('a')]!=0 and c<stack[-1]:
					top=stack.pop()
					visited[ord(top)-ord('a')]=0
				stack.append(c)
				visited[i]=1;
		for i in stack:
			string+=i
		return string

s="bcabc"
rd=RemoveDuplicateLetters()
print rd.removeDuplicateLetters(s)